File: frontend/PycpAstNode.py
from typing import Sequence, no_type_check
import enum, struct

def _pack_str(s: str) -> bytes:
	data = s.encode("utf-8")
	return struct.pack("I", len(data)) + data

def _pack_list(items: Sequence[PycpAstNode]) -> bytes:
	data = struct.pack("I", len(items))
	for item in items:
		data += item.as_bytecode()
	return data

class _Enum(enum.IntEnum):
	def __int__(self):
		return self.value

	def __str__(self) -> str:
		return self.name

	def __repr__(self) -> str:
		return self.name

class PycpAstNode:
	lineno: int | None = None

	class Type(_Enum):
		PROGRAM = enum.auto()

		STATEMENT = enum.auto()
		ASSIGNMENT_STATEMENT = enum.auto()
		RETURN_STATEMENT = enum.auto()
		EXPRESSION_STATEMENT = enum.auto()

		EXPRESSION = enum.auto()
		UNARY_EXPRESSION = enum.auto()
		BINARY_EXPRESSION = enum.auto()
		FUNCTION_EXPRESSION = enum.auto()
		CALL_EXPRESSION = enum.auto()
		IDENTIFIER_EXPRESSION = enum.auto()

		LITERAL = enum.auto()
		INTEGER_LITERAL = enum.auto()
		STRING_LITERAL = enum.auto()
		NONE_LITERAL = enum.auto()
			
	def _wrap_as_bytecode(self, node_type: int, payload: bytes = b"") -> bytes:
		lineno = self.lineno if self.lineno is not None else -1
		return (
						 struct.pack("H", node_type) +
						 struct.pack("i", lineno) +
						 struct.pack("I", len(payload)) +
						 payload
					 )

	def as_bytecode(self) -> bytes:
		return bytes()

	def __init__(self, lineno: int | None = None):
		self.lineno = lineno
	
	def __repr__(self) -> str:
		return "<Node>"

# 程序节点，即所有语句的集合
class Program(PycpAstNode):
	statements: list[Statement] = []

	def __init__(self, statements):
		super().__init__()
		self.statements = statements

	def as_bytecode(self) -> bytes:
		payload = _pack_list(self.statements)
		return self._wrap_as_bytecode(int(self.Type.PROGRAM), payload)

	def __repr__(self):
		return "\n".join(str(stmt) for stmt in self.statements)
	
	def append(self, statement: Statement) -> "Program":
		self.statements.append(statement)
		return self

# 语句节点
class Statement(PycpAstNode):
	def as_bytecode(self) -> bytes:
		return self._wrap_as_bytecode(int(self.Type.STATEMENT))

# 赋值语句 target = value
class AssignmentStatement(Statement):
	def __init__(self, target: str, value: "Expression", lineno=None):
		super().__init__(lineno)
		self.target = target
		self.value = value

	def __repr__(self):
		return f"<Assignment: {self.target} = {self.value}>"
	
	def as_bytecode(self) -> bytes:
		payload = _pack_str(self.target) + self.value.as_bytecode()
		return self._wrap_as_bytecode(int(self.Type.ASSIGNMENT_STATEMENT), payload)

# 返回语句
class ReturnStatement(Statement):
	def __init__(self, expression: "Expression", lineno=None):
		super().__init__(lineno)
		self.expression = expression

	def __repr__(self):
		return f"<Return: {self.expression}>"
	
	def as_bytecode(self) -> bytes:
		payload = self.expression.as_bytecode()
		return self._wrap_as_bytecode(int(self.Type.RETURN_STATEMENT),payload)

# 表达式语句，即单独的表达式作为语句
class ExpressionStatement(Statement):
	def __init__(self, expression: "Expression", lineno=None):
		super().__init__(lineno)
		self.expression = expression

	def __repr__(self):
		return f"<Expression: {self.expression}>"
	
	def as_bytecode(self) -> bytes:
		return self._wrap_as_bytecode(
																	int(self.Type.EXPRESSION_STATEMENT),
																	self.expression.as_bytecode()
																 )


# 表达式节点
class Expression(PycpAstNode):
	def as_bytecode(self) -> bytes:
		return self._wrap_as_bytecode(int(self.Type.EXPRESSION))

# 一元表达式
class UnaryExpression(Expression):
	# 一元运算支持的运算符
	class Operator(_Enum):
		UMINUS = enum.auto()

	def __init__(self, op: "Operator", operand: "Expression", lineno=None):
		super().__init__(lineno)
		self.op = op
		self.operand = operand

	def __repr__(self):
		return f"<Unary: {self.op} {self.operand}>"
	
	def as_bytecode(self) -> bytes:
		payload = struct.pack("H", int(self.op))  # 添加操作符
		payload += self.operand.as_bytecode()
		return self._wrap_as_bytecode(int(self.Type.UNARY_EXPRESSION), payload)

# 二元表达式
class BinaryExpression(Expression):
	# 二元运算支持的运算符
	class Operator(_Enum):
		PLUS = enum.auto()
		MINUS = enum.auto()
		MULTIPLY = enum.auto()
		DIVIDE = enum.auto()

	def __init__(self, op: "Operator", left: "Expression", right: "Expression", lineno=None):
		super().__init__(lineno)
		self.op = op
		self.left = left
		self.right = right

	def __repr__(self):
		return f"<Binary: {self.left} {self.op} {self.right}>"
	
	def as_bytecode(self) -> bytes:
		payload = struct.pack("H", int(self.op))  # 添加操作符！
		payload += self.left.as_bytecode()
		payload += self.right.as_bytecode()
		return self._wrap_as_bytecode(int(self.Type.BINARY_EXPRESSION), payload)

# 匿名函数表达式
class FunctionExpression(Expression):
	def __init__(self, params: list[str], body: Program, name: str = "@anonymous", lineno=None):
		super().__init__(lineno)
		self.params = params
		self.body = body
		self.name = name

	def __repr__(self):
		return f"<Function: {self.name}({self.params}) {self.body}>"
	
	def as_bytecode(self) -> bytes:
		payload = b""

		# 函数名
		payload += _pack_str(self.name)

		# 参数列表
		payload += struct.pack("I", len(self.params))
		for p in self.params:
				payload += _pack_str(p)

		# 函数体
		payload += self.body.as_bytecode()

		return self._wrap_as_bytecode(int(self.Type.FUNCTION_EXPRESSION), payload)

# 函数调用表达式
class CallExpression(Expression):
	def __init__(self, callee: "FunctionExpression", arguments: list["Expression"], lineno):
		super().__init__(lineno)
		self.callee = callee
		self.arguments = arguments

	def __repr__(self):
		return f"<Call: {self.callee}({self.arguments})>"
	
	def as_bytecode(self) -> bytes:
		payload = b""
		payload += self.callee.as_bytecode()

		payload += struct.pack("I", len(self.arguments))
		for arg in self.arguments:
			payload += arg.as_bytecode()

		return self._wrap_as_bytecode(int(self.Type.CALL_EXPRESSION), payload)

# 标识符表达式
class IdentifierExpression(Expression):
	def __init__(self, name, lineno=None):
		super().__init__(lineno)
		self.name = name

	def __repr__(self):
		return f"<Identifier: {self.name}>"
	
	def as_bytecode(self) -> bytes:
		payload = _pack_str(self.name)
		return self._wrap_as_bytecode(int(self.Type.IDENTIFIER_EXPRESSION),payload)


# 字面量
class Literal(Expression):
	def as_bytecode(self) -> bytes:
		return self._wrap_as_bytecode(int(self.Type.LITERAL))
		
# 整数字面量
class IntegerLiteral(Literal):
	def __init__(self, value: str, lineno=None):
		super().__init__(lineno)
		self.value = value

	def __repr__(self):
		return f"<Integer: {self.value}>"
	
	def as_bytecode(self) -> bytes:
		payload = _pack_str(self.value)
		return self._wrap_as_bytecode(int(self.Type.INTEGER_LITERAL), payload)
	
	def as_dict(self):
		return {
					   "type": self.Type.INTEGER_LITERAL, 
						 "value": self.value
					 }
	
class StringLiteral(Literal):
	def __init__(self, value, lineno=None):
		super().__init__(lineno)
		self.value = value

	def __repr__(self):
		return f"<String: \"{repr(self.value)}\">"
	
	def as_bytecode(self) -> bytes:
		payload = _pack_str(self.value)
		return self._wrap_as_bytecode(int(self.Type.STRING_LITERAL), payload)

class NoneLiteral(Literal):
	def __init__(self, lineno = None):
		super().__init__(lineno)

	def __repr__(self):
		return "<None>"
	
	def as_bytecode(self) -> bytes:
		return self._wrap_as_bytecode(int(self.Type.NONE_LITERAL))


class _ByteCodeReader:
	def __init__(self, data: bytes):
		self.data = data
		self.offset = 0

	def read(self, size: int) -> bytes:
		chunk = self.data[self.offset : self.offset + size]
		self.offset += size
		return chunk

	def read_u16(self) -> int:
		return struct.unpack("H", self.read(2))[0]

	def read_i32(self) -> int:
		return struct.unpack("i", self.read(4))[0]

	def read_u32(self) -> int:
		return struct.unpack("I", self.read(4))[0]

	def read_i64(self) -> int:
		return struct.unpack("q", self.read(8))[0]

	def read_string(self) -> str:
		length = self.read_u32()
		return self.read(length).decode("utf-8")

@no_type_check
def _read_node(reader: _ByteCodeReader) -> PycpAstNode:
	node_type = reader.read_u16()
	lineno = reader.read_i32()
	payload_len = reader.read_u32()

	payload_start = reader.offset

	node_type_enum = PycpAstNode.Type(node_type)

	
	match node_type_enum:

			case PycpAstNode.Type.PROGRAM:
					count = reader.read_u32()
					statements = [_read_node(reader) for _ in range(count)]
					node = Program(statements)

			case PycpAstNode.Type.ASSIGNMENT_STATEMENT:
					target = reader.read_string()
					value = _read_node(reader)
					node = AssignmentStatement(target, value) 

			case PycpAstNode.Type.RETURN_STATEMENT:
					expr = _read_node(reader)
					node = ReturnStatement(expr)

			case PycpAstNode.Type.EXPRESSION_STATEMENT:
					expr = _read_node(reader)
					node = ExpressionStatement(expr)

			case PycpAstNode.Type.UNARY_EXPRESSION:
					op = UnaryExpression.Operator(reader.read_u16())
					operand = _read_node(reader)
					node = UnaryExpression(op, operand)

			case PycpAstNode.Type.BINARY_EXPRESSION:
					op = BinaryExpression.Operator(reader.read_u16())
					left = _read_node(reader)
					right = _read_node(reader)
					node = BinaryExpression(op, left, right)

			case PycpAstNode.Type.FUNCTION_EXPRESSION:
					name = reader.read_string()
					param_count = reader.read_u32()
					params = [reader.read_string() for _ in range(param_count)]
					body = _read_node(reader)
					node = FunctionExpression(params, body, name=name)

			case PycpAstNode.Type.CALL_EXPRESSION:
					callee = _read_node(reader)
					arg_count = reader.read_u32()
					args = [_read_node(reader) for _ in range(arg_count)]
					node = CallExpression(callee, args, lineno)

			case PycpAstNode.Type.IDENTIFIER_EXPRESSION:
					name = reader.read_string()
					node = IdentifierExpression(name)

			case PycpAstNode.Type.INTEGER_LITERAL:
					value = reader.read_string()
					node = IntegerLiteral(value)

			case PycpAstNode.Type.STRING_LITERAL:
					value = reader.read_string()
					node = StringLiteral(value)

			case PycpAstNode.Type.NONE_LITERAL:
					node = NoneLiteral()

			case _:
					raise ValueError(f"Unknown node type: {node_type_enum}")

	node.lineno = None if lineno == -1 else lineno

	# 跳过剩余payload（安全校验）
	reader.offset = payload_start + payload_len

	return node

def from_bytecode(data: bytes) -> Program:
	reader = _ByteCodeReader(data)
	return _read_node(reader) # type: ignore

File: frontend/src/PycpAstNode.cpp
#include "PycpAstNode.hpp"

namespace Pycp::Ast {

// ============================================================
// 辅助函数
// ============================================================
static std::string unary_op_to_string(UnaryOp op) {
	switch (op) {
		case UnaryOp::UMINUS: return "UMINUS";
		default: return "UNKNOWN";
	}
}

static std::string binary_op_to_string(BinaryOp op) {
	switch (op) {
		case BinaryOp::PLUS:     			return "PLUS";
		case BinaryOp::MINUS:    			return "MINUS";
		case BinaryOp::MULTIPLY: 			return "MULTIPLY";
		case BinaryOp::DIVIDE:   			return "DIVIDE";
		case BinaryOp::POWER:    			return "POWER";
		case BinaryOp::LESS_THAN:     return "LESS_THAN";
		case BinaryOp::GREATER_THAN:  return "GREATER_THAN";
		case BinaryOp::LESS_EQUAL:    return "LESS_EQUAL";
		case BinaryOp::GREATER_EQUAL: return "GREATER_EQUAL";
		case BinaryOp::EQUAL:         return "EQUAL";
		case BinaryOp::NOT_EQUAL:     return "NOT_EQUAL";
		default: return "UNKNOWN";
	}
}

// ============================================================
// Program 实现
// ============================================================
Program::Program(std::vector<Statement*>* stmts)
	: statements(stmts) {}

Program::~Program() {
	if (statements == nullptr) {
	  return;
	}

	for (Statement* stmt : *statements) {
		delete stmt;
	}
	delete statements;
}

void Program::append(Statement* stmt) {
	statements->push_back(stmt);
}

std::string Program::to_string() const {
	std::string result;
	for (size_t i = 0; i < statements->size(); ++i) {
		if (i > 0) {
			result += "\n";
		}
		result += (*statements)[i]->to_string();
	}
	return result;
}

// ============================================================
// AssignmentStatement 实现
// ============================================================
AssignmentStatement::AssignmentStatement(Expression* t, Expression* v, int line)
	: target(t), value(v) {
	lineno = line;
}

AssignmentStatement::~AssignmentStatement() {
	delete value;
	delete target;
}

std::string AssignmentStatement::to_string() const {
	return "<Assignment: " + target->to_string() + " = " + value->to_string() + ">";
}

// ============================================================
// ReturnStatement 实现
// ============================================================
ReturnStatement::ReturnStatement(Expression* expr, int line)
	: expression(expr) {
	lineno = line;
}

ReturnStatement::~ReturnStatement() {
	delete expression;
}

std::string ReturnStatement::to_string() const {
	return "<Return: " + expression->to_string() + ">";
}

// ============================================================
// ExpressionStatement 实现
// ============================================================
ExpressionStatement::ExpressionStatement(Expression* expr, int line)
	: expression(expr) {
	lineno = line;
}

ExpressionStatement::~ExpressionStatement() {
	delete expression;
}

std::string ExpressionStatement::to_string() const {
	return "<Expression: " + expression->to_string() + ">";
}

// ============================================================
// UnaryExpression 实现
// ============================================================
UnaryExpression::UnaryExpression(UnaryOp o, Expression* operand_, int line)
	: op(o), operand(operand_) {
	lineno = line;
}

UnaryExpression::~UnaryExpression() {
	delete operand;
}

std::string UnaryExpression::to_string() const {
	return "<Unary: " + unary_op_to_string(op) + " " + operand->to_string() + ">";
}

// ============================================================
// BinaryExpression 实现
// ============================================================
BinaryExpression::BinaryExpression(BinaryOp o, Expression* l, Expression* r, int line)
	: op(o), left(l), right(r) {
	lineno = line;
}

BinaryExpression::~BinaryExpression() {
	delete left;
	delete right;
}

std::string BinaryExpression::to_string() const {
	return "<Binary: " + left->to_string() + " " + binary_op_to_string(op) + " " + right->to_string() + ">";
}

// ============================================================
// FunctionExpression 实现
// ============================================================
FunctionExpression::FunctionExpression(std::vector<std::string*> p,
                                       Program* b,
                                       std::string n, int line,
                                       Expression* deco)
	: name(std::move(n)), params(std::move(p)), body(b), decorator(deco) {
	lineno = line;
}

FunctionExpression::~FunctionExpression() {
	delete body;
	delete decorator;
	for (std::string* param : params) {
		delete param;
	}
}

std::string FunctionExpression::to_string() const {
	std::string params_str;
	for (size_t i = 0; i < params.size(); ++i) {
		if (i > 0) params_str += ", ";
		params_str += *(params[i]);
	}

	std::string body_str = body->to_string();
	// 缩进处理
	size_t pos = 0;
	while (pos < body_str.length()) {
		pos = body_str.find('\n', pos);
		if (pos == std::string::npos) break;
		body_str.insert(pos + 1, "  ");
		pos += 3;
	}
	if (!body_str.empty()) {
		body_str = "\n  " + body_str + "\n";
	}

	return "<Function: " + name + "(" + params_str + ")" + body_str + ">";
}

// ============================================================
// CallExpression 实现
// ============================================================
CallExpression::CallExpression(Expression* callee_,
                               std::vector<Expression*> args, int line)
	: callee(callee_), arguments(std::move(args)) {
	lineno = line;
}

CallExpression::~CallExpression() {
	delete callee;
	for (Expression* arg : arguments) {
		delete arg;
	}
}

std::string CallExpression::to_string() const {
	std::string args_str;
	for (size_t i = 0; i < arguments.size(); ++i) {
		if (i > 0) args_str += ", ";
		args_str += arguments[i]->to_string();
	}
	return "<Call: " + callee->to_string() + "(" + args_str + ")>";
}

// ============================================================
// IdentifierExpression 实现
// ============================================================
IdentifierExpression::IdentifierExpression(std::string* n, int line)
	: name(n) {
	lineno = line;
}

IdentifierExpression::~IdentifierExpression() {
	delete name;
}

std::string IdentifierExpression::to_string() const {
	return "<Identifier: " + *name + ">";
}

// ============================================================
// IntegerLiteral 实现
// ============================================================
IntegerLiteral::IntegerLiteral(std::string* v, int line)
	: value(v) {
	lineno = line;
}

IntegerLiteral::~IntegerLiteral() {
  delete value;
}

std::string IntegerLiteral::to_string() const {
	return "<Integer: " + *value + ">";
}

// ============================================================
// StringLiteral 实现
// ============================================================
StringLiteral::StringLiteral(std::string* v, int line)
	: value(v) {
	lineno = line;
}

StringLiteral::~StringLiteral() {
  delete value;
}

std::string StringLiteral::to_string() const {
	return "<String: \"" + *value + "\">";
}

// ============================================================
// NoneLiteral 实现
// ============================================================
NoneLiteral::NoneLiteral(int line) {
	lineno = line;
}

std::string NoneLiteral::to_string() const {
	return "<None>";
}

// ============================================================
// IfBranch 实现
// ============================================================
IfBranch::IfBranch(Expression* cond, Program* b, int line, bool elif)
	: condition(cond), body(b), is_elif(elif) {
	lineno = line;
}

IfBranch::~IfBranch() {
	delete condition;
	delete body;
}

// 对代码块内部做缩进处理，与 FunctionExpression 保持一致
static std::string indent_block(const std::string& raw, const std::string& base_indent) {
	std::string body_str = raw;
	size_t pos = 0;
	while (pos < body_str.length()) {
		pos = body_str.find('\n', pos);
		if (pos == std::string::npos) break;
		body_str.insert(pos + 1, base_indent + "  ");
		pos += base_indent.length() + 3;
	}
	if (!body_str.empty()) {
		body_str = "\n" + base_indent + "  " + body_str + "\n" + base_indent;
	}
	return body_str;
}

std::string IfBranch::to_string() const {
	return to_string("");
}

std::string IfBranch::to_string(const std::string& indent) const {
	std::string body_str = indent_block(body->to_string(), indent);

	if (condition == nullptr) {
		// else 分支：无条件
		return indent + "<Else:" + body_str + ">";
	}
	const std::string keyword = is_elif ? "<Elif: " : "<If: ";
	return indent + keyword + condition->to_string() + body_str + ">";
}

// ============================================================
// IfStatement 实现
// ============================================================
IfStatement::IfStatement(IfBranch* if_br,
                         std::vector<IfBranch*>* elif_brs,
                         Program* else_b,
                         int line)
	: if_branch(if_br), elif_branches(elif_brs), else_body(else_b) {
	lineno = line;
}

IfStatement::~IfStatement() {
	delete if_branch;
	if (elif_branches != nullptr) {
		for (IfBranch* br : *elif_branches) {
			delete br;
		}
		delete elif_branches;
	}
	delete else_body;
}

std::string IfStatement::to_string() const {
	// 不显式包裹组合节点标签，仅以 If/Elif/Else 区分各分支并体现层级关联。
	const std::string indent = "  ";
	std::string result = if_branch->to_string(indent);

	if (elif_branches != nullptr) {
		for (IfBranch* br : *elif_branches) {
			result += "\n" + br->to_string(indent);
		}
	}

	if (else_body != nullptr) {
		// else 分支以独立 IfBranch（condition==nullptr）形式复用打印逻辑
		IfBranch else_branch(nullptr, else_body);
		result += "\n" + else_branch.to_string(indent);
	}

	return result;
}

// ============================================================
// ImportStatement 实现
// ============================================================
ImportStatement::ImportStatement(std::string* mod, std::string* al, int line)
	: module_name(mod), alias(al) {
	lineno = line;
}

ImportStatement::~ImportStatement() {
	delete module_name;
	delete alias;
}

std::string ImportStatement::to_string() const {
	if (alias != nullptr) {
		return "<Import: " + *module_name + " as " + *alias + ">";
	}
	return "<Import: " + *module_name + ">";
}

// ============================================================
// FromImportStatement 实现
// ============================================================
FromImportStatement::FromImportStatement(std::string* mod,
                                         std::vector<std::string*>* ns, int line)
	: module_name(mod), names(ns) {
	lineno = line;
}

FromImportStatement::~FromImportStatement() {
	delete module_name;
	if (names != nullptr) {
		for (std::string* n : *names) delete n;
		delete names;
	}
}

std::string FromImportStatement::to_string() const {
	std::string result = "<FromImport: " + *module_name + " import ";
	if (names != nullptr) {
		for (std::size_t i = 0; i < names->size(); ++i) {
			if (i > 0) result += ", ";
			result += *((*names)[i]);
		}
	}
	return result + ">";
}

// ============================================================
// AttributeExpression 实现
// ============================================================
AttributeExpression::AttributeExpression(Expression* t, std::string* a, int line)
	: target(t), attr(a) {
	lineno = line;
}

AttributeExpression::~AttributeExpression() {
	delete target;
	delete attr;
}

std::string AttributeExpression::to_string() const {
	return "<Attribute: " + target->to_string() + "." + *attr + ">";
}

// ============================================================
// ListLiteral 实现
// ============================================================
ListLiteral::ListLiteral(std::vector<Expression*>* elems, int line)
	: elements(elems) {
	lineno = line;
}

ListLiteral::~ListLiteral() {
	if (elements != nullptr) {
		for (Expression* e : *elements) delete e;
		delete elements;
	}
}

std::string ListLiteral::to_string() const {
	std::string result = "[";
	if (elements != nullptr) {
		for (std::size_t i = 0; i < elements->size(); ++i) {
			if (i > 0) result += ", ";
			result += (*elements)[i]->to_string();
		}
	}
	return result + "]";
}

// ============================================================
// IndexExpression 实现
// ============================================================
IndexExpression::IndexExpression(Expression* t, Expression* idx, int line)
	: target(t), index(idx) {
	lineno = line;
}

IndexExpression::~IndexExpression() {
	delete target;
	delete index;
}

std::string IndexExpression::to_string() const {
	return target->to_string() + "[" + index->to_string() + "]";
}

// ============================================================
// ClassDefinition 实现
// ============================================================
ClassDefinition::ClassDefinition(std::string* n,
                                 std::string* parent,
                                 std::vector<Statement*>* mv,
                                 std::vector<Statement*>* ms, int line)
	: name(n), parent_name(parent), member_variables(mv), methods(ms) {
	lineno = line;
}

ClassDefinition::~ClassDefinition() {
	delete name;
	delete parent_name;
	if (member_variables != nullptr) {
		for (Statement* s : *member_variables) delete s;
		delete member_variables;
	}
	if (methods != nullptr) {
		for (Statement* s : *methods) delete s;
		delete methods;
	}
}

std::string ClassDefinition::to_string() const {
	std::string result = "<Class: " + (name ? *name : "?");
	if (parent_name != nullptr) {
		result += " inherits " + *parent_name;
	}
	if (member_variables != nullptr) {
		for (Statement* s : *member_variables) {
			result += "\n  " + s->to_string();
		}
	}
	if (methods != nullptr) {
		for (Statement* s : *methods) {
			result += "\n  " + s->to_string();
		}
	}
	return result + ">";
}

// ============================================================
// MemberVariable 实现
// ============================================================
MemberVariable::MemberVariable(std::string* n, Expression* v, int line,
                               Expression* decorator)
	: name(n), value(v), decorator(decorator) {
	lineno = line;
}

MemberVariable::~MemberVariable() {
	delete name;
	delete value;
	delete decorator;
}

std::string MemberVariable::to_string() const {
	std::string dec = (decorator != nullptr) ? ("@" + decorator->to_string() + " ") : "";
	if (value != nullptr) {
		return "<Member: " + dec + (name ? *name : "?") + " = " + value->to_string() + ">";
	}
	return "<Member: " + dec + (name ? *name : "?") + ">";
}

// ============================================================
// MethodDefinition 实现
// ============================================================
MethodDefinition::MethodDefinition(FunctionExpression* f, int line,
                                   Expression* decorator)
	: function(f), decorator(decorator) {
	lineno = line;
}

MethodDefinition::~MethodDefinition() {
	delete function;
	delete decorator;
}

std::string MethodDefinition::to_string() const {
	std::string dec = (decorator != nullptr) ? ("@" + decorator->to_string() + " ") : "";
	return "<Method: " + dec + (function ? function->to_string() : "?") + ">";
}

// ============================================================
// RepeatStatement 实现
// ============================================================
RepeatStatement::RepeatStatement(RepeatMode m,
                                 Expression* cnt, Expression* st,
                                 Expression* en, Expression* step,
                                 Expression* cond, std::string* var,
                                 Program* b, int line)
	: mode(m), count_expr(cnt), start_expr(st), end_expr(en),
	  step_expr(step), cond_expr(cond), var_name(var), body(b) {
	lineno = line;
}

RepeatStatement::~RepeatStatement() {
	delete count_expr;
	delete start_expr;
	delete end_expr;
	delete step_expr;
	delete cond_expr;
	delete var_name;
	delete body;
}

std::string RepeatStatement::to_string() const {
	const char* kind = "?";
	switch (mode) {
		case RepeatMode::INFINITE: kind = "infinite"; break;
		case RepeatMode::WHILE:    kind = "while";    break;
		case RepeatMode::COUNT:    kind = "count";    break;
		case RepeatMode::RANGE:    kind = "range";    break;
	}
	std::string var = (var_name != nullptr) ? (" as " + *var_name) : "";
	return "<Repeat " + std::string(kind) + var + ">";
}

// ============================================================
// BreakStatement 实现
// ============================================================
BreakStatement::BreakStatement(int line)
	: Statement() {
	lineno = line;
}

BreakStatement::~BreakStatement() = default;

std::string BreakStatement::to_string() const {
	return "<Break>";
}

// ============================================================
// ForeachStatement 实现
// ============================================================
ForeachStatement::ForeachStatement(std::string* var, Expression* it,
                                   Program* b, int line)
	: var_name(var), iterable(it), body(b) {
	lineno = line;
}

ForeachStatement::~ForeachStatement() {
	delete var_name;
	delete iterable;
	delete body;
}

std::string ForeachStatement::to_string() const {
	std::string var = (var_name != nullptr) ? *var_name : "?";
	std::string it = (iterable != nullptr) ? iterable->to_string() : "?";
	return "<For " + var + " in " + it + ">";
}

// ============================================================
// ClassExpression 实现
// ============================================================
ClassExpression::ClassExpression(std::string* parent,
                                 std::vector<Statement*>* mv,
                                 std::vector<Statement*>* ms, int line)
	: parent_name(parent), member_variables(mv), methods(ms) {
	lineno = line;
}

ClassExpression::~ClassExpression() {
	delete parent_name;
	if (member_variables != nullptr) {
		for (Statement* s : *member_variables) delete s;
		delete member_variables;
	}
	if (methods != nullptr) {
		for (Statement* s : *methods) delete s;
		delete methods;
	}
}

std::string ClassExpression::to_string() const {
	std::string result = "<ClassExpr: ?";
	if (parent_name != nullptr) result += " inherits " + *parent_name;
	if (member_variables != nullptr) {
		for (Statement* s : *member_variables) result += "\n  " + s->to_string();
	}
	if (methods != nullptr) {
		for (Statement* s : *methods) result += "\n  " + s->to_string();
	}
	return result + ">";
}

} // namespace Pycp::AstNode